src/esp/gfx/DebugLineRender.cpp

```diff
@@ -0,0 +1,314 @@
+    CORRADE_INTERNAL_ASSERT(lerpFraction >= 0.f && lerpFraction <= 1.f);
+    Mn::Vector3 clippedPt = Mn::Math::lerp(*pt1, *pt0, lerpFraction);
+    *pt1 = clippedPt;
+    return true;
+  }
+
+  return true;
+}
+
+}  // namespace
+
+DebugLineRender::DebugLineRender()
+    : _glResources{std::make_unique<GLResourceSet>()} {
+  _glResources->mesh.addVertexBuffer(_glResources->buffer, 0,
+                                     Mn::Shaders::VertexColor3D::Position{},
+                                     Mn::Shaders::VertexColor3D::Color4{});
+}
+
+void DebugLineRender::releaseGLResources() {
+  _glResources = nullptr;
+}
+
+void DebugLineRender::drawLine(const Mn::Vector3& from,
+                               const Mn::Vector3& to,
+                               const Mn::Color4& fromColor,
+                               const Mn::Color4& toColor) {
+  VertexRecord v1{from, remapAlpha(fromColor)};
+  VertexRecord v2{to, remapAlpha(toColor)};
+  arrayAppend(_verts, {v1, v2});
+}
+
+void DebugLineRender::setLineWidth(float lineWidth) {
+  // This is derived from experiments with glLineWidth on Nvidia hardware.
+  const float maxLineWidth = 20.f;
+  if (lineWidth > maxLineWidth) {
+    LOG(WARNING) << "DebugLineRender::setLineWidth: requested lineWidth of "
+                 << lineWidth
+                 << " is greater than "
+                    "max supported width of "
+                 << maxLineWidth;
+    lineWidth = maxLineWidth;
+  }
+  _internalLineWidth = lineWidth / 2;  // see also DebugLineRender::flushLines
+}
+
+void DebugLineRender::flushLines(const Magnum::Matrix4& camMatrix,
+                                 const Magnum::Matrix4& projMatrix,
+                                 const Magnum::Vector2i& viewport) {
+  CORRADE_ASSERT(_glResources,
+                 "DebugLineRender::flushLines: no GL resources; see "
+                 "also releaseGLResources", );
+
+  bool doToggleBlend = !glIsEnabled(GL_BLEND);
+
+  if (doToggleBlend) {
+    Mn::GL::Renderer::enable(Mn::GL::Renderer::Feature::Blending);
+    Mn::GL::Renderer::setBlendFunction(
+        Mn::GL::Renderer::BlendFunction::SourceAlpha,
+        Mn::GL::Renderer::BlendFunction::OneMinusSourceAlpha);
+    Mn::GL::Renderer::setBlendEquation(Mn::GL::Renderer::BlendEquation::Add);
+  }
+
+  // Update buffer with new data
+  _glResources->buffer.setData(_verts, Mn::GL::BufferUsage::DynamicDraw);
+
+  // Update shader
+  _glResources->mesh.setCount(_verts.size());
+
+  glDisable(GL_LINE_SMOOTH);  // anti-aliased lines cause artifacts
+
+  glLineWidth(_internalLineWidth);
+
+  // We draw lines multiple times, with pixel offsets, to produce a single
+  // thick, visually-appealing line. Todo: implement thick lines using
+  // triangles, for example https://mattdesl.svbtle.com/drawing-lines-is-hard.
+  // 1.2 is hand-tuned for Nvidia hardware to be just small enough so we don't
+  // see gaps between the individual offset lines.
```

Comment:
To be clear: the current thick-line implementation here is sloppy in the sense of not producing perfectly-rendered thick lines. We can upgrade in the future without changing the public API, if we wish.